### PR TITLE
Platooning control plugin handles control switchover

### PR DESCRIPTION
**Pull request by TonysCousin**
# PR Details
## Description
Enhanced the platooning control plugin to cooperate with other control plugins to accommodate the fact that all control plugins publish their output on a common ros topic, and this topic is only intended to carry data from a single controller at a time.  When trajectory executor switches from one control plugin to another, we want the old one to stop publishing.  However, it needs to wait some amount of time since it is spinning faster than the producer of its input trajectory data. Since this input data is not guaranteed to arrive at a precise interval, the control plugin needs to wait for a conservative timeout period before deciding that the input stream has dried up.

Note: architecture documentation is to be updated at a later time.

## Related Issue

PR #1647 

## Motivation and Context

Ensures a control plugin stops publishing output data when another control plugin takes over.

## How Has This Been Tested?

Not yet tested - this PR is for integration testing.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: platooning_control/src/platoon_control.cpp [hunk 1]
```diff
@@ -99,6 +103,22 @@ namespace platoon_control
 
     bool PlatoonControlPlugin::controlTimerCb()
     {
+        // If it has been a long time since input data has arrived then reset the input counter and return
+        // Note: this quiets the controller after its input stream stops, which is necessary to allow 
+        // the replacement controller to publish on the same output topic after this one is done.
+        long current_time = ros::Time::now().toNsec() / 1000000;
+        if (current_time - prev_input_time_ > config_.shutdownTimeout)
```
Comment (saina-ramyar) [inline platooning_control/src/platoon_control.cpp:110]: I suggest we add some log statements to see what is value of "current_time - prev_input_time_" so we can tune the parameter

File: platooning_control/src/platoon_control.cpp [hunk 2]
```diff
@@ -99,6 +103,22 @@ namespace platoon_control
...
+        {
+            consecutive_input_counter_ = 0;
+            return;
+        }
+
+        // If there have not been enough consecutive timely inputs then return (waiting for
+        // previous control plugin to time out and stop publishing, since it uses same output topic)
+        if (consecutive_input_counter_ <= config_.ignoreInitialInputs)
+        {
+            return;
```
Comment (saina-ramyar) [inline platooning_control/src/platoon_control.cpp:120]: again, a log statement could be useful

File: platooning_control/src/platoon_control.cpp [hunk 3]
```diff
@@ -109,17 +129,17 @@ namespace platoon_control
...
-        
+    void  PlatoonControlPlugin::trajectoryPlan_cb(const cav_msgs::TrajectoryPlan::ConstPtr& tp)
+    {
         if (tp->trajectory_points.size() < 2) {
             ROS_WARN_STREAM("PlatoonControlPlugin cannot execute trajectory as only 1 point was provided");
             return;
         }
 
         latest_trajectory_ = *tp;
+        prev_input_time_ = ros::Time::now().toNsec() / 1000000;
```
Comment (saina-ramyar) [inline platooning_control/src/platoon_control.cpp:141]: again Log statemet saying not received a new trajectory plan
Also, can you add log statements printing "prev_input_time_" and "tp->header.stamp.toNsec() / 1000000"
I think its better to use the trajectory plan's own timestamp, but I remember those didn't have reliable values in the past, so want to make sure.

File: platooning_control/src/platoon_control.cpp (hunk 1 above)
Comment (TonysCousin) [reply platooning_control/src/platoon_control.cpp:110]: Good idea.

File: platooning_control/src/platoon_control.cpp (hunk 2 above)
Comment (TonysCousin) [reply platooning_control/src/platoon_control.cpp:120]: Yep

File: platooning_control/src/platoon_control.cpp (hunk 3 above)
Comment (TonysCousin) [reply platooning_control/src/platoon_control.cpp:141]: Sure.

**Pull request merged**